synapse/handlers/e2e_keys.py

```diff
@@ -175,6 +175,33 @@ async def query_devices(
                     user_ids_not_in_cache,
                     remote_results,
                 ) = await self.store.get_user_devices_from_cache(query_list)
+
+                # Check that the homeserver still shares a room with all cached users.
+                # Note that this check may be slightly racy when a remote user leaves a
+                # room after we have fetched their cached device list. In the worst case
+                # we will do extra federation queries for devices that we had cached.
+                cached_users = set(remote_results.keys())
+                valid_cached_users = (
+                    await self.store.get_users_server_still_shares_room_with(
+                        remote_results.keys()
+                    )
+                )
+                invalid_cached_users = cached_users - valid_cached_users
+                if invalid_cached_users:
+                    # Fix up results. If we get here there may be bugs in device list
+                    # tracking.
+                    # TODO: is this actually useful? it doesn't fix the case where a
+                    #       remote user rejoins a room and we query their device list
+                    #       after.
+                    user_ids_not_in_cache.update(invalid_cached_users)
+                    for invalid_user_id in invalid_cached_users:
+                        remote_results.pop(invalid_user_id)
+                    logger.error(
+                        "Devices for %s were cached, but the server no longer shares "
+                        "any rooms with them. The cached device lists are stale.",
+                        invalid_cached_users,
+                    )
+
```

Comment:
I'm not fully convinced this check is useful, because it'll only catch `/keys/query` requests done while we no longer share a room with a remote user, which I wouldn't expect to happen often, or at all.

It's also slightly racy and can trigger spuriously when a remote user leaves a room while we're in the middle of handling the request.

Do we want to keep it or remove it?

Reply:
I think its valid for clients to do a `/keys/query` for users they don't share a room with, e.g. because they're about to start a DM or something?

Reply:
That sounds plausible!